flink-runtime/src/main/java/org/apache/flink/runtime/watermark/SourceWatermark.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.watermark;
+
+import java.io.Serializable;
+
+/**
+ * This represents the watermark for a single source partition.
+ */
+public class SourceWatermark implements Serializable {
+
+	private static final long serialVersionUID = 1L;
+	private long timestamp;
```

Comment:
What does the timestamp represent? Is it when the watermark last changed or when it was last communicated by the subtask (even if it did not change, for example because the subtask is just reading a lot of data under the same watermark). We will need a way to detect that a source subtask is idle so we can avoid waiting for it (similar to how we has to identify idle within a subtask).

Reply:
The timestamp here is meant to represent the watermark itself -- the current low watermark for the sub-task that sent it.

I do agree, however, that we will also need to know at what time the watermark was sent so that we can ignore it if it hasn't been updated in some configurable amount of time.

Very good point.